src/pages/Admin/Room/Edit/RoomEdit.tsx

```diff
@@ -134,24 +54,96 @@ const EditRoom: React.FC<EditRoomProps> = ({
 
       <Modal.Body>
         <Form onSubmit={handleSubmit(onSubmit)}>
-          {renderNameInput()}
-          {renderUrlInput()}
-          {renderDescriptionInput()}
-          {renderLogoInput()}
-
-          {roomTemplate?.customInputs?.map((input) => renderCustomInput(input))}
-
-          {renderTemplateSelect()}
+          <Form.Row>
+            <div className="input-wrapper">
+              <Form.Label>Name your room</Form.Label>
+              <Form.Control
+                type="text"
+                ref={register}
+                name="title"
+                placeholder="Room name"
+                custom
+              />
+            </div>
+          </Form.Row>
+
+          <Form.Row>
+            <div className="input-wrapper">
+              <Form.Label>The room url</Form.Label>
+              <Form.Control
+                type="text"
+                ref={register}
+                name="url"
+                placeholder="Room url"
+                custom
+              />
+            </div>
+          </Form.Row>
+
+          <Form.Row>
+            <div className="input-wrapper">
+              <Form.Label>The room description (optional)</Form.Label>
+              <Form.Control
+                type="text"
+                ref={register}
+                name="description"
+                placeholder="Description"
+                custom
+              />
+            </div>
+          </Form.Row>
+
+          <Form.Row>
+            <div className="input-wrapper">
+              <Form.Label>
+                How you want the room to appear on the map
+              </Form.Label>
+              <ImageInput
+                onChange={handleImageChange}
+                name="image"
+                forwardRef={register}
+                small
+                nameWithUnderscore
+                imgUrl={room.image_url}
+              />
+            </div>
+          </Form.Row>
+
+          {roomTemplate?.customInputs?.map((input) => (
```

Comment:
Potentially this `.map` should be `useMemo`'d to prevent needing to be run on every render of this component.

Reply:
I'm trying to fix this and I get this warning which is preventing me from committing to this branch.
```  Line 62:5:  React Hook useMemo has a missing dependency: 'register'. Either include it or remove the dependency array      react-hooks/exhaustive-deps```

the code looks like this:

```
const customInputFields = useMemo(() =>
    roomTemplate?.customInputs?.map((input) => (
      <div className="room-edit-modal__input" key={input.name}>
        <span>{input.title}</span>
        <input type="text" name={input.name} ref={register} />
      </div>
    )
  ),
    [roomTemplate]
  );
```

I only half understand how hooks work. If you have any good articles/pages to read, could you share?

Reply:
Got a similar error for the useCallback one too:

```  Line 50:6:  React Hook useCallback has a missing dependency: 'setValue'. Either include it or remove the dependency array  react-hooks/exhaustive-deps```

with the code 

```
const handleImageChange = useCallback((val: string) => {
    setValue("image_url", val, false);
}, [room.image_url]);
```

Reply:
For general info on hooks, I recommend starting at https://reactjs.org/docs/hooks-intro.html and reading through the 8 or so related items. You can skip/skim through bits if it's not immediately relevant.

Those warnings relate to the 'rules of hooks':

- https://reactjs.org/docs/hooks-rules.html

It relates to the 'deps' (dependencies) array at the end of the hook. While that exists on a few different hooks, this section seems to explain it for the `useEffect` hook (it works the same way on all of them):

- https://reactjs.org/docs/hooks-effect.html#tip-optimizing-performance-by-skipping-effects

Basically, while the value in the 'deps' remain the same, the hook will only run once. When any of the 'deps' change, it will 'break' this 'memoisation' ('break memo'), and run again.

The warning/error you're seeing here is basically saying that there is some variable/thing used within the 'body' of the hook, that isn't included within that 'deps' array. Which means that the hook won't properly re-run  all the times its meant to (which can lead to weird/hard to find/debug issues)

Reply:
Generally I don't actually type the values into there manually, as my IDE allows me to automatically apply the fix, and fill the array for me.

eg. 
![image](https://user-images.githubusercontent.com/753891/108975160-4e16d780-76da-11eb-8e45-55bc47398a68.png)

When I click on the blue ESLint 'fix', it will do it for me. I believe you can probably also apply the fix via the command line, but most IDE's should have similar functionality, and doing it in the IDE is way quicker/easier.

Reply:
Good to know, thanks! I think I've fixed everything up now.

Reply:
Awesome :)